config/credential_handler.rb

```diff
@@ -0,0 +1,151 @@
+    when Doorkeeper::OAuth::CLIENT_CREDENTIALS
+      # This flow does not require an authorization code but instead requires JWT validation using JWK sets.
+      # This is not functionality supported by Doorkeeper intrinsically, so the custom validation is implemented here.
+      # Once the validation handled here passes, the client id and secret are sent along and no further validation is
+      # required before Doorkeeper sends an access token for this flow.
+      handle_client_credentials_flow(request)
+    else
+      # Unsupported grant type.
+      # NOTE: It should be impossible to get here with Doorkeeper's initial validation.
+      raise Doorkeeper::Errors::InvalidTokenStrategy
+    end
+  end
+
+  # Sara Alert supports the Smart on FHIR Backend Services protocol
+  # as described here: https://hl7.org/fhir/uv/bulkdata/authorization/index.html
+  # This requires the use client_credentials grant flow with JWT assertions.
+  # This method handles authorizing the client by decoding the JWT assertion for the supported client credentials flow. 
+  def self.handle_client_credentials_flow(request)
+    # Extract the shared client ID.
+    client_id = request.parameters[:client_id]
+    raise_missing_param_error('client_id') unless client_id.present?
+
+    # Extract signed JWT assertion from client_assertion param.
+    # This is what is signed with the client secret and will be used to authorize this request.
+    signed_token = request.parameters[:client_assertion]
+    raise_missing_param_error('client_assertion') unless signed_token.present?
+
+    # Find the associated client application. 
+    # Return if not found - Doorkeeper will automatically throw invalid_client error in this case.
+    client_application = Doorkeeper::Application.find_by(uid: client_id)
+    return unless client_application.present?
+    
+    # Find the registered public key set for this client application to decode the JWT assertion.
+    if client_application[:public_key_set].present?
+      # Public key set is a stored serialized object that needs to be loaded first.
+      public_key_set = HashWithIndifferentAccess.new(SafeYAML.load(client_application[:public_key_set]))
+      raise_invalid_JWK_error unless public_key_set.present? && public_key_set[:keys].present?
+    else
+      raise_invalid_JWK_error
+    end
+
+    # Get expected JWT aud value: the request token endpoint.
+    aud = request.original_url
+
+    # Begin process of decoding and validating JWT assertion required for this flow.
+    begin
+      # Use JWT library for decoding and validation.
+      decoded_token = JWT.decode(
+        signed_token,
+        nil,
+        true,
+        { 
+          algorithms: ['RS384'],
+          jwks: public_key_set,
+          iss: client_id,
+          verify_iss: true,
+          sub: client_id, 
+          verify_sub: true,
+          aud: aud,
+          verify_aud: true,
+          verify_iat: true,
```

Comment:
The `verify_iat` field doesn't seem to be effecting my testing.

Reply:
Yeah me neither - and it's not required in the protocol. I removed it.